### ISSUE TITLE
Sort notes using locale sorting rules

### ISSUE THREAD

**Issue by chylex** [stefan-niedermann/nextcloud-notes#1504]
**Is your feature request related to a problem? Please describe.**
Hi! The Notes app has localization for the Czech language, but it does not use Czech sorting rules. For example, if I have recipes and title my notes by the names of dishes, in the Android app I see them ordered like this:
```
* Svíčková
* Zelí
* Špagety <--
```

According to the Czech sorting rules, `Š` follows immediately after `S`, so the order should be:
```
* Svíčková
* Špagety <--
* Zelí
```

**Describe the solution you'd like**
It seems the app automatically detects my system language and has localized texts, so it should also use locale-based sorting based on the detected language. It's confusing to have an app that's localized, but then uses English sorting rules.

It would be nice to also have the option to change the language instead of auto-detecting it, say for people who want to see English and have English sorting rules even though their system languge is different, but that's a separate issue.

**Comment by stefan-niedermann** [comment]
Hi,

thanks for the report, that's an interesting case which I didn't stumble upon yet. I can confirm, that this als happens in german (my localized language):

![IMG_20220421_074532](https://user-images.githubusercontent.com/4741199/164382827-c499a9ee-dbca-43c0-b0f1-0f7f7525a33e.jpg)

Of course we don't implement the sorting ourself but rely on the database to do the heavy work for us.

[Relevant source code](https://github.com/stefan-niedermann/nextcloud-notes/blob/master/app/src/main/java/it/niedermann/owncloud/notes/persistence/dao/NoteDao.java#L42)

Basically:

```sql
ORDER BY title COLLATE NOCASE ASC
```

We will for sure not implement this in Java for performance reasons, but If someone know how to handle locale sorting with SQLite databases, let me hear 🤷

**Comment by korelstar** [comment]
According to https://stackoverflow.com/questions/19906086/problems-ordering-sqlite-by-a-column-with-accented-characters-%C3%81 you should try
```
ORDER BY column COLLATE LOCALIZED
```

**Comment by stefan-niedermann** [comment]
@korelstar while that basically works, `LOCALIZED` seems to not taking care about upper / lower case.

![image](https://user-images.githubusercontent.com/4741199/164397063-90a6a839-87f0-4374-a8be-9021912c8672.png)

Unfortunately I am not that deep in SQL, but do you know a way to combine the `COLLATE` statements `LOCALIZED` and (the currently used) `NOCASE`?

**Comment by stefan-niedermann** [comment]
Well, just checked the currently released app version and `NOCASE` also doesn't seem to handle this situation today, looks more like an "ignore case"...

**Comment by korelstar** [comment]
> Well, just checked the currently released app version and `NOCASE` also doesn't seem to handle this situation today, looks more like an "ignore case"...

I think that is exactly the purpose of `NOCASE`. And isn't this what we want? So if `LOCALIZED` ignores the case, too, that is exactly what we want! Not?

**Comment by IngoBelka** [comment]
What about something like this:
`SELECT City FROM Cities ORDER BY City COLLATE UNICODE`

**Comment by chylex** [comment]
Does the database know the locale it should use for sorting? Just want to make sure it's handled properly, because it's not possible to come up with a set of rules that would work for all languages.

For example in Czech, words beginning with 'ch' immediately follow words beginning with 'h', even though 'ch' are just two separate characters. A similar thing happens in Slovak, where the alphabet goes "d, ď, dz, dž".

**Comment by stefan-niedermann** [comment]
> I think that is exactly the purpose of NOCASE. And isn't this what we want? So if LOCALIZED ignores the case, too, that is exactly what we want! Not?

Yes, you are right.

> What about something like this

`UNICODE` will work for many cases, but `LOCALIZED` fits better to what we want to achieve and does not necessarily match the order of `UNICODE`

> Does the database know the locale it should use for sorting? 

Yes it does. Android passes the `Context` to the database, so it is aware of the current locale.

I'll provide a dev build this weekend sk everyone can test the changed implementation and provide feedback 🙂

**Comment by stefan-niedermann** [comment]
@IngoBelka @chylex as promised, [here you can find the dev build](https://github.com/stefan-niedermann/nextcloud-notes/suites/6230567984/artifacts/220985527) - Some feedback for different languages would be great :slightly_smiling_face: (Unzip the file and install the `.apk` - it is safe to install it in parallel to the released Play Store and / or F-Droid versions)

**Comment by chylex** [comment]
Quick note, when I logged into the dev app, it said "Importing 16 out of 15 notes" :D not sure what happened, my notes folder has 15 files. It's also laggy when scrolling on the page with all my 15 notes, not sure if that's just an issue in the dev build, but scrolling in the official build is very smooth.

The sorting for Czech seems to work correctly, thanks.

**Comment by stefan-niedermann** [comment]
Thanks for the feedback!

> Importing 16 out of 15 notes

Yep, noticed that too, but it's only a visual glitch and independent from this issue or the current changes 😉 

>  It's also laggy when scrolling on the page with all my 15 notes

Yeah, it's only a quick debug build without any optimizations. This won't happen in a release build.

The changes will land in `3.4.20`, I'll just try to add an unit test before I merge the stuff.

**Comment by stefan-niedermann** [comment]
Fix will be published in `3.5.0`.
